tools/build-version.go

```diff
@@ -36,15 +36,11 @@ func getTag(match ...string) (string, *semver.PRVersion) {
 }
 
 func main() {
-	tags, err := exec.Command("git", "tag").Output()
-	if err != nil {
-		log.Println("git tag:", err)
-	}
-	if len(tags) == 0 {
-		// no tags found -- fetch them
-		err = exec.Command("git", "fetch", "--tags").Run()
-		if err != nil {
-			log.Println("git fetch --tags:", err)
+	// Fetch the latest tags
+	if err := exec.Command("git", "fetch", "--tags", "--unshallow").Run(); err != nil {
+		// unshallow fails locally, so try it again
+		if err := exec.Command("git", "fetch", "--tags").Run(); err != nil {
+			log.Println("git fetch --tags", err)
```

Comment:
Do you have an example this log output? I guess this output doesn't include the actual output (stderr) of the `git fetch --tags` command?

Reply:
`git fetch --tags` doesn't fail, it's `git describe --tags --match v*` which "fails" (nothing found) and due to this `semver.ParseTolerant: strconv.ParseUint: parsing "": invalid syntax` happens:
https://github.com/zyedidia/micro/actions/runs/9547202303/job/26311706304#step:4:46

This was the reason why I pushed that much debug stuff to this PR, to find out why it fails.

> Why does it fail?

As far as I've understood GitHub will by default and with `--depth=1` create a repository copy in shallow mode.
An alternative to my proposal could be, to apply `fetch-depth: 0` (https://github.com/actions/checkout), but this will clone the whole repository every time within the action. The latter I didn't test yet.

Reply:
In general I don't like it with these two fetch commands either. An alternative is given and needs to be tested. Then we've to decide, what's has smallest impact here.

Reply:
I meant: why does unshallowing fail?
Ok, now I got it: it fails if the repo is already not shallow.

Looks like it can checked more directly via `git rev-parse --is-shallow-repository`, but your current version seems fine too. (Perhaps just make the comment more clear, e.g. "unshallow fails if the repo is already not shallow"?)

...And what about my first question? In case if `git fetch --tags` fails, as I can see, all we print is:

```
2024/06/18 14:54:05 git fetch --tags exit status 1
```

so we don't know why it fails. So, as I noted earlier, we might also want to print its stderr to know more:

```
if e, ok := err.(*exec.ExitError); ok {
	log.Print(string(e.Stderr))
}
```

Reply:
> An alternative to my proposal could be, to apply `fetch-depth: 0` (https://github.com/actions/checkout), but this will clone the whole repository every time within the action.

We fetch the whole (unshallowed) repo later anyway, right?

Hmm, perhaps we could try playing with `git ls-remote --tags` to avoid fetching the whole repo. (But maybe it isn't worth it.)

Reply:
> so we don't know why it fails. So, as I noted earlier, we might also want to print its stderr to know more:

I've overseen that hint. Sorry!

> We fetch the whole (unshallowed) repo later anyway, right?

Thought it's was related to the tags only, but the docs say it too...it's then fully fetched as non-shallow.
So from my perspective we can then directly clone with `--depth=0` resp. `fetch-depth: 0`, since the main reason is that `git describe --tags` will fail.